File: Runtime/TransitionAttributeTarget.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.StochasticTime

Friend Class TransitionAttributeTarget
    Inherits STSimDistributionBase

    Private m_TransitionAttributeTargetId As Integer
    Private m_TransitionAttributeTypeId As Integer
    Private m_TargetRemaining As Double
    Private m_ExpectedAmount As Double
    Private m_Multiplier As Double = 1.0

    Public Sub New(
        ByVal transitionAttributeTargetId As Integer,
        ByVal iteration As Nullable(Of Integer),
        ByVal timestep As Nullable(Of Integer),
        ByVal stratumId As Nullable(Of Integer),
        ByVal secondaryStratumId As Nullable(Of Integer),
        ByVal tertiaryStratumId As Nullable(Of Integer),
        ByVal transitionAttributeTypeId As Integer,
        ByVal targetAmount As Nullable(Of Double),
        ByVal distributionTypeId As Nullable(Of Integer),
        ByVal distributionFrequency As Nullable(Of DistributionFrequency),
        ByVal distributionSD As Nullable(Of Double),
        ByVal distributionMin As Nullable(Of Double),
        ByVal distributionMax As Nullable(Of Double))

        MyBase.New(
            iteration,
            timestep,
            stratumId,
            secondaryStratumId,
            tertiaryStratumId,
            targetAmount,
            distributionTypeId,
            distributionFrequency,
            distributionSD,
            distributionMin,
            distributionMax)

        Me.m_TransitionAttributeTargetId = transitionAttributeTargetId
        Me.m_TransitionAttributeTypeId = transitionAttributeTypeId

    End Sub

    Public ReadOnly Property TransitionAttributeTargetId As Integer
        Get
            Return Me.m_TransitionAttributeTargetId
        End Get
    End Property

    Public ReadOnly Property TransitionAttributeTypeId As Integer
        Get
            Return Me.m_TransitionAttributeTypeId
        End Get
    End Property

    Public Property TargetRemaining As Double
        Get
            Return Me.m_TargetRemaining
        End Get
        Set(value As Double)
            Me.m_TargetRemaining = value
        End Set
    End Property

    Public Property Multiplier As Double
        Get
            Return Me.m_Multiplier
        End Get
        Set(ByVal value As Double)
            Me.m_Multiplier = value
        End Set
    End Property

    Public Property ExpectedAmount As Double
        Get
            Return Me.m_ExpectedAmount
        End Get
        Set(ByVal value As Double)
            Me.m_ExpectedAmount = value
        End Set
    End Property

    Public Overrides Function Clone() As STSimDistributionBase

        Dim t As New TransitionAttributeTarget(
            Me.TransitionAttributeTargetId,
            Me.Iteration,
            Me.Timestep,
            Me.StratumId,
            Me.SecondaryStratumId,
            Me.TertiaryStratumId,
            Me.TransitionAttributeTypeId,
            Me.DistributionValue,
            Me.DistributionTypeId,
            Me.DistributionFrequency,
            Me.DistributionSD,
            Me.DistributionMin,
            Me.DistributionMax)

        t.TargetRemaining = Me.TargetRemaining
        t.Multiplier = Me.Multiplier
        t.ExpectedAmount = Me.ExpectedAmount

        Return t

    End Function

End Class

File: Runtime/TransitionAttributeTarget.cs
// A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2018 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using SyncroSim.Common;
using SyncroSim.StochasticTime;
using System.Diagnostics;
using System.Collections.Generic;

namespace SyncroSim.STSim
{
    internal class TransitionAttributeTarget : STSimDistributionBase
    {
        private int m_TransitionAttributeTargetId;
        private int m_TransitionAttributeTypeId;
        private double m_TargetRemaining;
        private double m_ExpectedAmount;
        private double m_Multiplier = 1.0;
        private List<TransitionAttributeTargetPrioritization> m_Prioritizations;
        private MultiLevelKeyMap5<TransitionAttributeTargetPrioritization> m_PrioritizationMap;
        private TransitionAttributeTargetPrioritization m_DefaultPrioritization;

        public TransitionAttributeTarget(
            int transitionAttributeTargetId, int? iteration, int? timestep, int? stratumId, int? secondaryStratumId, int? tertiaryStratumId, 
            int transitionAttributeTypeId, double? targetAmount, int? distributionTypeId, DistributionFrequency? distributionFrequency, 
            double? distributionSD, double? distributionMin, double? distributionMax) : base(iteration, timestep, stratumId, secondaryStratumId,
                tertiaryStratumId, targetAmount, distributionTypeId, distributionFrequency, distributionSD, distributionMin, distributionMax)
        {
            this.m_TransitionAttributeTargetId = transitionAttributeTargetId;
            this.m_TransitionAttributeTypeId = transitionAttributeTypeId;
        }

        public int TransitionAttributeTargetId
        {
            get
            {
                return this.m_TransitionAttributeTargetId;
            }
        }

        public int TransitionAttributeTypeId
        {
            get
            {
                return this.m_TransitionAttributeTypeId;
            }
        }

        public double TargetRemaining
        {
            get
            {
                this.CHECK_DISABLED();
                return this.m_TargetRemaining;
            }
            set
            {
                this.CHECK_DISABLED();
                this.m_TargetRemaining = value;
            }
        }

        public double TargetRemainingNoCheck
        {
            get
            {
                return this.m_TargetRemaining;
            }
        }

        public double Multiplier
        {
            get
            {
                this.CHECK_DISABLED();
                return this.m_Multiplier;
            }
            set
            {
                this.CHECK_DISABLED();
                this.m_Multiplier = value;
            }
        }

        public double MultiplierNoCheck
        {
            get
            {
                return this.m_Multiplier;
            }
        }

        public double ExpectedAmount
        {
            get
            {
                this.CHECK_DISABLED();
                return this.m_ExpectedAmount;
            }
            set
            {
                this.CHECK_DISABLED();
                this.m_ExpectedAmount = value;
            }
        }

        public double ExpectedAmountNoCheck
        {
            get
            {
                return this.m_ExpectedAmount;
            }
        }

        public List<TransitionAttributeTargetPrioritization> Prioritizations
        {
            get
            {
                return m_Prioritizations;
            }
        }

        internal TransitionAttributeTargetPrioritization DefaultPrioritization
        {
            get
            {
                return m_DefaultPrioritization;
            }

            set
            {
                m_DefaultPrioritization = value;
            }
        }

        public override STSimDistributionBase Clone()
        {
            TransitionAttributeTarget t = new TransitionAttributeTarget(
                this.TransitionAttributeTargetId, this.Iteration, this.Timestep, this.StratumId, this.SecondaryStratumId, this.TertiaryStratumId, 
                this.TransitionAttributeTypeId, this.DistributionValue, this.DistributionTypeId, this.DistributionFrequency, this.DistributionSD, 
                this.DistributionMin, this.DistributionMax);

            t.TargetRemaining = this.TargetRemainingNoCheck;
            t.Multiplier = this.MultiplierNoCheck;
            t.ExpectedAmount = this.ExpectedAmountNoCheck;
            t.IsDisabled = this.IsDisabled;

            t.SetPrioritizations(this.Prioritizations);
            t.DefaultPrioritization = this.DefaultPrioritization;

            return t;
        }

        public void SetPrioritizations(List<TransitionAttributeTargetPrioritization> prioritizations)
        {
            this.ClonePrioritizationList(prioritizations);
            this.CreatePrioritizationMap();
        }

        public TransitionAttributeTargetPrioritization GetPrioritization(
            int stratumId,
            int? secondaryStratumId,
            int? tertiaryStratumId,
            int? transitionGroupId,
            int stateClassId)
        {
            //Look for a prioritization in the map.  If it is not found, return the default.

            TransitionAttributeTargetPrioritization pri = this.m_PrioritizationMap.GetItem(
                stratumId, secondaryStratumId, tertiaryStratumId, transitionGroupId, stateClassId);

            if (pri == null)
            {
                pri = this.m_DefaultPrioritization;
            }

            return pri;
        }

        private void ClonePrioritizationList(List<TransitionAttributeTargetPrioritization> prioritizations)
        {
            Debug.Assert(this.m_Prioritizations == null);
            this.m_Prioritizations = new List<TransitionAttributeTargetPrioritization>();

            foreach (TransitionAttributeTargetPrioritization t in prioritizations)
            {
                this.m_Prioritizations.Add(new TransitionAttributeTargetPrioritization(
                    t.Iteration,
                    t.Timestep,
                    t.TransitionAttributeTypeId,
                    t.StratumId,
                    t.SecondaryStratumId,
                    t.TertiaryStratumId,
                    t.TransitionGroupId,
                    t.StateClassId,
                    t.Priority));
            }

            Debug.Assert(this.m_Prioritizations.Count == prioritizations.Count);
        }

        private void CreatePrioritizationMap()
        {
            Debug.Assert(this.m_PrioritizationMap == null);
            this.m_PrioritizationMap = new MultiLevelKeyMap5<TransitionAttributeTargetPrioritization>();

            foreach (TransitionAttributeTargetPrioritization pri in this.m_Prioritizations)
            {
                TransitionAttributeTargetPrioritization p = this.m_PrioritizationMap.GetItemExact(
                    pri.StratumId, pri.SecondaryStratumId, pri.TertiaryStratumId, pri.TransitionGroupId, pri.StateClassId);

                if (p == null)
                {
                    this.m_PrioritizationMap.AddItem(
                        pri.StratumId,
                        pri.SecondaryStratumId,
                        pri.TertiaryStratumId,
                        pri.TransitionGroupId,
                        pri.StateClassId,
                        pri);
                }
            }
        }
    }
}
